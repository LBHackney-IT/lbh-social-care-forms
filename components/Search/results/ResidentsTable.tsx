import Link from 'next/link';
import cx from 'classnames';
import { LegacyResident, User } from 'types';
import { canManageCases } from '../../../lib/permissions';
import { useAuth } from '../../UserContext/UserContext';
import styles from './ResidentsTable.module.scss';

const ResultEntry = ({
  person,
  newTab,
}: {
  person: LegacyResident;
  newTab?: boolean;
}): React.ReactElement => {
  const { user } = useAuth() as { user: User };
  const { mosaicId, firstName, lastName, dateOfBirth, address, ageContext } =
    person;

  const isRecordRestricted = !canManageCases(user, {
    contextFlag: ageContext,
    restricted: person.restricted,
  });

  return (
    <tr
      className={cx(
        'govuk-table__row',
        isRecordRestricted && styles.restrictedRow
      )}
    >
      <td className="govuk-table__cell">{mosaicId}</td>
      <td className="govuk-table__cell">
        {newTab ? (
          <a
            href={`/residents/${mosaicId}`}
            target="_blank"
            rel="noreferrer"
            className="govuk-link govuk-custom-text-color"
          >
            {firstName} {lastName}
          </a>
        ) : (
          <Link href={`/residents/${mosaicId}`}>
            <a className="govuk-link govuk-custom-text-color">
              {firstName} {lastName}
            </a>
          </Link>
        )}
      </td>
      <td className="govuk-table__cell">
        {dateOfBirth && new Date(dateOfBirth).toLocaleDateString('en-GB')}
      </td>
      <td className="govuk-table__cell">
        {address && (
          <>
            {address?.address && (
              <span>
                {address.address.length > 20
                  ? address?.address.substr(0, 19).trimEnd() + '... '
                  : address?.address + ' ' || ''}
              </span>
            )}
            <span className={styles.uppercase}>
              {(address && address.postcode) || ''}
            </span>
          </>
        )}
      </td>
      <td className="govuk-table__cell govuk-table__cell--numeric">
        {isRecordRestricted && (
          <span
            className={cx('govuk-tag lbh-tag lbh-tag--grey', styles.uppercase)}
          >
            Restricted
          </span>
        )}
      </td>
    </tr>
  );
};

const ResultTable = ({
  records,
  newTab,
}: {
  records: LegacyResident[];
  /** whether to open residents in a new tab? */
  newTab?: boolean;
}): React.ReactElement => (
  <table className="govuk-table lbh-table" data-testid="residents-table">
    <thead className="govuk-table__head">
      <tr className="govuk-table__row">
        <th scope="col" className="govuk-table__header">
          Social care ID
        </th>
        <th scope="col" className="govuk-table__header">
          Person
        </th>
        <th scope="col" className="govuk-table__header">
          Date of birth
        </th>
        <th scope="col" className="govuk-table__header">
          Address
        </th>
        <th scope="col" className="govuk-table__header" />
      </tr>
    </thead>
    <tbody className="govuk-table__body">
      {records.map((result) => (
        <ResultEntry key={result.mosaicId} person={result} newTab={newTab} />
      ))}
    </tbody>
  </table>
);

export default ResultTable;
