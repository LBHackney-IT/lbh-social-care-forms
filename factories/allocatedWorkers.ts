import { Factory } from 'fishery';

import { Allocation } from 'types';

export const allocationFactory = Factory.define<Allocation>(({ sequence }) => ({
  id: sequence,
  allocatedWorkerTeam: 'Safeguarding and Reviewing Team',
  allocatedWorker: 'Officer Name',
  allocationEndDate: '2019-03-28 00:00:00',
  allocationStartDate: '2019-03-28 00:00:00',
  workerType: 'Consultant Social Worker',
  personId: 1234,
  caseStatus: 'Open',
  personName: 'foo',
  personAddress: 'the address',
  personDateOfBirth: '2020-03-20',
  ragRating: 'amber',
}));

export const mockedAllocation = allocationFactory.build();
export const mockedAllocations = [allocationFactory.build()];
