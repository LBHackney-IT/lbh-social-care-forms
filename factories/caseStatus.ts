import { Factory } from 'fishery';
import {
  CaseStatus,
  CaseStatusFields,
  AddCaseStatusFormData,
  CaseStatusAnswerDisplay,
} from 'types';

export const mockedCaseStatusFactory = Factory.define<CaseStatus>(
  ({ sequence }) => ({
    id: sequence,
    type: 'CIN',
    answers: [mockedStatusField.build()],
    startDate: '2021-01-01T02:00:00Z',
    endDate: '2021-12-01T02:00:00Z',
    notes: 'a lot to note',
  })
);

export const mockedCaseStatusAddRequest = Factory.define<AddCaseStatusFormData>(
  ({ sequence }) => ({
    personId: sequence,
    type: 'CIN',
    fields: [{ name: 'test', selected: 'test' }],
    startDate: '2021-01-01T02:00:00Z',
    endDate: '2021-12-01T02:00:00Z',
    notes: 'a lot to note',
    createdby: 'test@testington.com',
  })
);

export const mockedStatusField = Factory.define<CaseStatusFields>(() => ({
  option: 'placementType',
  value: 'K1',
  startDate: '2020-08-01',
  createdAt: '2021-10-11T10:54:32Z',
}));

export const mockedCaseStatusAnswers = Factory.define<CaseStatusAnswerDisplay>(
  () => ({
    startDate: '2021-09-10',
    endDate: '2021-10-10',
    status: [mockedStatusField.build()],
  })
);
