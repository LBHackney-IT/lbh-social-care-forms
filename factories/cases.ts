import { Factory } from 'fishery';

import { Case } from 'types';

export const caseFactory = Factory.define<Case>(({ sequence }) => ({
  recordId: sequence.toString(),
  personId: 123,
  dateOfEvent: '25/10/2020 13:49:43',
  formName: 'foorm',
  officerEmail: 'Fname.Lname@hackney.gov.uk',
  caseFormData: {
    note: 'I am the note',
    first_name: 'Foo',
    last_name: 'Bar',
    timestamp: '11/03/2021 16:51:56',
    mosaic_id: 123,
    form_name_overall: 'foo',
    worker_email: 'worker@hackney.gov.uk',
    form_name: 'Foo bar',
    context_flag: 'A',
    date_of_event: '25/10/2020 13:49:43',
  },
  deleted: false,
  isImported: false,
}));

export const mockedNote = caseFactory.build({ caseFormUrl: 'https://foo.bar' });
export const mockedAllocationNote = caseFactory.build({
  formName: 'Worker allocated',
  caseFormData: {
    allocation_id: 321,
    created_by: 'foo@hackney.gov.uk',
    form_name_overall: 'API_Allocation',
    form_name: 'Worker allocated',
  },
});
export const mockedDeallocationNote = caseFactory.build({
  formName: 'Worker allocated',
  caseFormData: {
    allocation_id: 321,
    created_by: 'foo@hackney.gov.uk',
    deallocation_reason: 'deallocated because',
    form_name_overall: 'API_Deallocation',
    form_name: 'Worker deallocated',
  },
});
export const mockedCaseNote = caseFactory.build({
  caseFormData: {
    form_name_overall: 'ASC_case_note',
    case_note_title: 'i am a case title',
  },
});
export const mockedDeletedCaseNote = caseFactory.build({
  caseFormData: {
    form_name_overall: 'ASC_case_note',
    case_note_title: 'i am a case title',
  },
  formName: 'adult-case-note',
  deleted: true,
});
export const mockedWarningNoteCase = caseFactory.build({
  formName: 'Warning Note',
  caseFormData: {
    form_name_overall: 'API_WarningNote',
    warning_note_id: 456,
  },
});
