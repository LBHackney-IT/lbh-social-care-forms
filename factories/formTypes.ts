import { Factory } from 'fishery';
import { FormValue, FormStatusValue } from 'types';

export const mockedFormValueFactory = Factory.define<FormValue>(() => ({
  name: 'test',
  description: 'description',
  options: [mockedFormStatusValueFactory.build()],
}));

export const mockedFormStatusValueFactory = Factory.define<FormStatusValue>(
  () => ({
    name: 'N1',
    description: 'Noo 1',
  })
);

export const mockedFormValue = [mockedFormValueFactory.build()];
