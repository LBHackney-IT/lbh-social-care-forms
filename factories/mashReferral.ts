import { Factory } from 'fishery';
import { MashReferral, ReferralStage } from 'types';
import { mockedWorker } from './workers';

export const mashReferralFactory = Factory.define<MashReferral>(
  ({ sequence }) => ({
    id: sequence.toString(),
    referrer: 'hardcoded-referrer',
    requestedSupport: 'test-requested-support',
    assignedTo: mockedWorker,
    createdAt: '2021-10-27T09:32:17.319Z',
    clients: ['test-client'],
    referralDocumentURI: 'test-referral-document-URI',
    stage: ReferralStage.INITIAL,
    contactUrgentContactRequired: false,
    contactCreatedAt: '2021-10-28T10:55:45.097Z',
    initialReferralCategory: 'test-referral-category',
    initialDecision: 'test-initial-decision',
    initialUrgentContactRequired: false,
    initialCreatedAt: '2021-11-01T15:12:05.097Z',
    screeningDecision: 'test-screening-decision',
    screeningUrgentContactRequired: false,
    screeningCreatedAt: '2021-11-03T18:52:05.097Z',
    finalDecision: 'test-final-decision',
    finalReferralCategory: 'test-final-referral-category',
    finalUrgentContactRequired: false,
    finalCreatedAt: '2021-11-18T15:10:05.097Z',
  })
);

export const mockedMashReferral = mashReferralFactory.build();
export const mockedMashReferrals = [mashReferralFactory.build()];
