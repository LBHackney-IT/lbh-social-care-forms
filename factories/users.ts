import { Factory } from 'fishery';

import { User } from 'types';

export const userFactory = Factory.define<User>(() => ({
  name: 'foo',
  hasAdminPermissions: true,
  hasChildrenPermissions: true,
  hasAdultPermissions: true,
  hasDevPermissions: false,
  isAuditable: false,
  email: 'foo@bar.com',
  permissionFlag: 'A',
  isAuthorised: true,
  isInWorkflowsPilot: false,
  isInSafeguardingReviewing: false,
  isInPlacementManagementUnit: false,
}));

export const mockedOnlyAdultUser = userFactory.build({
  hasAdminPermissions: false,
  hasChildrenPermissions: false,
  hasDevPermissions: false,
});

export const mockedOnlyChildUser = userFactory.build({
  hasAdminPermissions: false,
  hasAdultPermissions: false,
  hasDevPermissions: false,
});

export const mockedAdminUser = userFactory.build({
  hasDevPermissions: false,
});

export const mockedUser = userFactory.build();

export const mockedUserInWorkflowsPilot = userFactory.build({
  isInWorkflowsPilot: true,
});
