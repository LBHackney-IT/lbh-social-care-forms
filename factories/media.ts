import { Media } from 'types';
import { mockedResident } from 'factories/residents';
import { mockedWorker } from 'factories/workers';
import { mockSubmission } from './submissions';

export const mockedMedia: Media = {
  mediaId: 'abc123',
  title: 'Example media file',
  resident: mockedResident,
  uploadedBy: mockedWorker,
  uploadedAt: '2021-06-21T12:00:00.000Z',
  submission: mockSubmission,
  files: {
    original: {
      type: 'image/png',
      size: 20000000,
      url: 'http://placehold.it/3000x2000',
    },
    medium: {
      type: 'image/png',
      size: 200000,
      url: 'http://placehold.it/500x333',
    },
    thumbnail: {
      type: 'image/png',
      size: 20000,
      url: 'http://placehold.it/100x150',
    },
  },
};
