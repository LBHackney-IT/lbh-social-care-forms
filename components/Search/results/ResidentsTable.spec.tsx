import { render } from '@testing-library/react';

import ResidentsTable from './ResidentsTable';
import {
  legacyResidentFactory,
  mockedLegacyResident,
} from 'factories/residents';
import { UserContext } from '../../UserContext/UserContext';
import { mockedUser, userFactory } from '../../../factories/users';

describe('ResidentsTable component', () => {
  it('should render a list of residents', () => {
    const { asFragment } = render(
      <UserContext.Provider value={{ user: mockedUser }}>
        <ResidentsTable records={[mockedLegacyResident]} />
      </UserContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render a restricted flag against any residents that the current user cannot manage cases for', () => {
    const { getByText } = render(
      <UserContext.Provider
        value={{
          user: userFactory.build({
            hasAdminPermissions: false,
            hasDevPermissions: false,
            hasChildrenPermissions: false,
            hasUnrestrictedPermissions: false,
            hasAllocationsPermissions: false,
            hasAdultPermissions: true,
          }),
        }}
      >
        <ResidentsTable
          records={[
            legacyResidentFactory.build({
              ageContext: 'C',
            }),
          ]}
        />
      </UserContext.Provider>
    );

    getByText('Restricted');
  });

  it('should render a resident without an address', () => {
    const { asFragment } = render(
      <UserContext.Provider
        value={{
          user: userFactory.build({
            hasAdminPermissions: false,
            hasDevPermissions: false,
            hasChildrenPermissions: false,
            hasUnrestrictedPermissions: false,
            hasAllocationsPermissions: false,
            hasAdultPermissions: true,
          }),
        }}
      >
        <ResidentsTable
          records={[
            legacyResidentFactory.build({
              ageContext: 'C',
              address: undefined,
            }),
          ]}
        />
      </UserContext.Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('when the address is less than 20 characters should render the full address and postcode', () => {
    const { asFragment } = render(
      <UserContext.Provider
        value={{
          user: userFactory.build({
            hasAdminPermissions: false,
            hasDevPermissions: false,
            hasChildrenPermissions: false,
            hasUnrestrictedPermissions: false,
            hasAllocationsPermissions: false,
            hasAdultPermissions: true,
          }),
        }}
      >
        <ResidentsTable
          records={[
            legacyResidentFactory.build({
              ageContext: 'C',
              address: {
                address: 'short address',
                postcode: 'E8 1AA',
              },
            }),
          ]}
        />
      </UserContext.Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('when the address is more than 20 characters should render the truncated address and postcode', () => {
    const { asFragment } = render(
      <UserContext.Provider
        value={{
          user: userFactory.build({
            hasAdminPermissions: false,
            hasDevPermissions: false,
            hasChildrenPermissions: false,
            hasUnrestrictedPermissions: false,
            hasAllocationsPermissions: false,
            hasAdultPermissions: true,
          }),
        }}
      >
        <ResidentsTable
          records={[
            legacyResidentFactory.build({
              ageContext: 'C',
              address: {
                address: 'really really long address',
                postcode: 'E8 1AA',
              },
            }),
          ]}
        />
      </UserContext.Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
