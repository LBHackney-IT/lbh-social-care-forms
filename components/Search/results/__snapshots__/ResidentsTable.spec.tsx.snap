// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResidentsTable component should render a list of residents 1`] = `
<DocumentFragment>
  <table
    class="govuk-table lbh-table"
    data-testid="residents-table"
  >
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__header"
          scope="col"
        >
          Social care ID
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Person
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Date of birth
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Address
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        />
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          1
        </td>
        <td
          class="govuk-table__cell"
        >
          <a
            class="govuk-link govuk-custom-text-color"
            href="/residents/1"
          >
            Foo Bar
          </a>
        </td>
        <td
          class="govuk-table__cell"
        >
          13/11/2020
        </td>
        <td
          class="govuk-table__cell"
        >
          <span>
            sjakdjlk 
          </span>
          <span
            class="uppercase"
          >
            hdsadjk
          </span>
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        />
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`ResidentsTable component should render a resident without an address 1`] = `
<DocumentFragment>
  <table
    class="govuk-table lbh-table"
    data-testid="residents-table"
  >
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__header"
          scope="col"
        >
          Social care ID
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Person
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Date of birth
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Address
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        />
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row restrictedRow"
      >
        <td
          class="govuk-table__cell"
        >
          3
        </td>
        <td
          class="govuk-table__cell"
        >
          <a
            class="govuk-link govuk-custom-text-color"
            href="/residents/3"
          >
            Foo Bar
          </a>
        </td>
        <td
          class="govuk-table__cell"
        >
          13/11/2020
        </td>
        <td
          class="govuk-table__cell"
        />
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          <span
            class="govuk-tag lbh-tag lbh-tag--grey uppercase"
          >
            Restricted
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`ResidentsTable component when the address is less than 20 characters should render the full address and postcode 1`] = `
<DocumentFragment>
  <table
    class="govuk-table lbh-table"
    data-testid="residents-table"
  >
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__header"
          scope="col"
        >
          Social care ID
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Person
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Date of birth
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Address
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        />
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row restrictedRow"
      >
        <td
          class="govuk-table__cell"
        >
          4
        </td>
        <td
          class="govuk-table__cell"
        >
          <a
            class="govuk-link govuk-custom-text-color"
            href="/residents/4"
          >
            Foo Bar
          </a>
        </td>
        <td
          class="govuk-table__cell"
        >
          13/11/2020
        </td>
        <td
          class="govuk-table__cell"
        >
          <span>
            short address 
          </span>
          <span
            class="uppercase"
          >
            E8 1AA
          </span>
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          <span
            class="govuk-tag lbh-tag lbh-tag--grey uppercase"
          >
            Restricted
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`ResidentsTable component when the address is more than 20 characters should render the truncated address and postcode 1`] = `
<DocumentFragment>
  <table
    class="govuk-table lbh-table"
    data-testid="residents-table"
  >
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__header"
          scope="col"
        >
          Social care ID
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Person
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Date of birth
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Address
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        />
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row restrictedRow"
      >
        <td
          class="govuk-table__cell"
        >
          5
        </td>
        <td
          class="govuk-table__cell"
        >
          <a
            class="govuk-link govuk-custom-text-color"
            href="/residents/5"
          >
            Foo Bar
          </a>
        </td>
        <td
          class="govuk-table__cell"
        >
          13/11/2020
        </td>
        <td
          class="govuk-table__cell"
        >
          <span>
            really really long... 
          </span>
          <span
            class="uppercase"
          >
            E8 1AA
          </span>
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          <span
            class="govuk-tag lbh-tag lbh-tag--grey uppercase"
          >
            Restricted
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
